### DISCUSSION TITLE
defaultValues are overwritten in useFieldArray

### DISCUSSION THREAD

**Discussion by ryanmk54** [react-hook-form/react-hook-form#7710] [Q&A] [answered]
I have code similar to the following https://codesandbox.io/s/youthful-leakey-uzvs0?file=/src/App.tsx
using @mui/material. My local project overwrites defaultValues when I change a value in useFieldArray. I tried adding Object.freeze to help debug the issue. It throws an error in my local copy, but not the codesandbox. I got the following stack trace.
utils.js:335 Uncaught TypeError: Cannot assign to read only property 'amt' of object '#<Object>'
    at set (index.esm.mjs:459)
    at updateValidAndValue (index.esm.mjs:1717)
    at Object.register (index.esm.mjs:2435)

set:
`    object[key] = newValue;`
updateValidAndValue:
`
      isUndefined(defaultValue) || ref && ref.defaultChecked || shouldSkipSetValueAs ? set(_formValues, name, shouldSkipSetValueAs ? defaultValue : getFieldValue(field._f)) : setFieldValue(name, defaultValue);
`
register
`    field ? disabledIsDefined && set(_formValues, name, options.disabled ? undefined : get(_formValues, name, getFieldValue(field._f))) : updateValidAndValue(name, true, options.value);
`

I am wondering if it could be my babel config, just because I don't why it would overwrite the object locally and not in the code sandbox. I've attached [babel.config.js.txt](https://github.com/react-hook-form/react-hook-form/files/7982613/babel.config.js.txt) just in case. Is there any reason why it would modify defaultValues?

**Comment by bluebill1049** [comment]
as far as I know, we only update `defautValues` with those reset actions. Are you reset some individual fields or the entire form?

**Reply by ryanmk54** [reply]
I commented out all my calls to reset and the problem persists.

Even without changing the form, the page crashes inside Object.register, if I do Object.freeze(defaultValues.details[0]).

Some interesting information I found, when the form is first rendered:
defaultValues.details[0] === getValues('details')[0] // true
defaultValues.details[0] === fieldArray.fields[0] // false

When I update a field, it updates defaultValues.details[0], but not fieldArray.fields[0]

**Comment by ryanmk54** [comment, marked as answer]
I think I found a bug in the [cloneObject method](https://github.com/react-hook-form/react-hook-form/blob/master/src/utils/cloneObject.ts#L16).
`
      if (isFunction(data[key])) {
        copy = data;
        break;
      }
`
Should this be `copy[key] = data[key]`?

The bug is shown by adding functions to the Array.prototype. If I delete the added properties from the prototype, everything works.
